vsintegration/src/FSharp.Editor/Refactor/AddExplicitReturnType.fs

```diff
@@ -0,0 +1,125 @@
+﻿namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System.Composition
+open System.Threading
+open System.Threading.Tasks
+open FSharp.Compiler.CodeAnalysis
+open FSharp.Compiler.Symbols
+open FSharp.Compiler.Text
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Text
+open Microsoft.CodeAnalysis.CodeRefactorings
+open Microsoft.CodeAnalysis.CodeActions
+open CancellableTasks
+
+[<ExportCodeRefactoringProvider(FSharpConstants.FSharpLanguageName, Name = "AddExplicitReturnType"); Shared>]
+type internal AddExplicitReturnType [<ImportingConstructor>] () =
+    inherit CodeRefactoringProvider()
+
+    static member isValidMethodWithoutTypeAnnotation
+        (symbolUse: FSharpSymbolUse)
+        (parseFileResults: FSharpParseFileResults)
+        (funcOrValue: FSharpMemberOrFunctionOrValue)
+        =
+        let typeAnnotationRange =
+            parseFileResults.TryRangeOfReturnTypeHint(symbolUse.Range.Start, false)
+
+        let isLambdaIfFunction =
+            funcOrValue.IsFunction
+            && parseFileResults.IsBindingALambdaAtPosition symbolUse.Range.Start
+
+        let res =
+            (not funcOrValue.IsValue || not isLambdaIfFunction)
+            && not (funcOrValue.ReturnParameter.Type.IsUnresolved)
+            && not (parseFileResults.IsTypeAnnotationGivenAtPosition symbolUse.Range.Start)
+            && not typeAnnotationRange.IsNone
+
+        match res with
+        | true -> Some funcOrValue
+        | false -> None
+
+    static member refactor
+        (context: CodeRefactoringContext)
+        (symbolUse: FSharpSymbolUse, memberFunc: FSharpMemberOrFunctionOrValue, parseFileResults: FSharpParseFileResults)
+        =
+        let title = SR.AddExplicitReturnTypeAnnotation()
+
+        let getChangedText (sourceText: SourceText) =
+            let inferredType = memberFunc.ReturnParameter.Type.TypeDefinition.DisplayName
+
+            let rangeOfReturnType =
+                parseFileResults.TryRangeOfReturnTypeHint(symbolUse.Range.Start, false)
+
+            let textChange =
+                rangeOfReturnType
+                |> Option.map (fun range -> RoslynHelpers.FSharpRangeToTextSpan(sourceText, range))
+                |> Option.map (fun textSpan -> TextChange(textSpan, $":{inferredType}"))
```

Comment:
So I think we should actually align to the [style guide](https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting) which has a [section](https://learn.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-return-type-annotations) about return type annotations prescribing extra spaces before and _after_ the colon.

Sorry I missed that earlier, that would need adjusting the tests a bit as well :(

Reply:
No Problem. Literally a few spaces :)